### PR TITLE
Merged latest yolo version

### PR DESCRIPTION
**Pull request by bongmo**
Merged latest yolo version

### CODE REVIEW

**Comment by yjxiong** [comment]
Thanks! Will look at it.

On Thursday, February 18, 2016, Bong Mo Kim notifications@github.com
wrote:

> ## Merged latest yolo version
> 
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/yjxiong/darknet/pull/1
> Commit Summary
> - pydarknet for latest version
> - Fix readme
> - Fix readme
> 
> File Changes
> - _M_ CMakeLists.txt
>   https://github.com/yjxiong/darknet/pull/1/files#diff-0 (22)
> - _M_ Makefile https://github.com/yjxiong/darknet/pull/1/files#diff-1
>   (6)
> - _M_ README.md
>   https://github.com/yjxiong/darknet/pull/1/files#diff-2 (2)
> - _M_ cfg/darknet.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-3 (1)
> - _A_ cfg/extraction.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-4 (190)
> - _A_ cfg/extraction.conv.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-5 (179)
> - _A_ cfg/msr_152.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-6 (1598)
> - _A_ cfg/msr_34.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-7 (366)
> - _A_ cfg/msr_50.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-8 (558)
> - _A_ cfg/rnn.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-9 (40)
> - _A_ cfg/rnn.train.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-10 (40)
> - _M_ cfg/strided.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-11 (12)
> - _A_ cfg/yolo-coco.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-12 (240)
> - _M_ cfg/yolo-small.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-13 (123)
> - _A_ cfg/yolo-tiny.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-14 (140)
> - _M_ cfg/yolo.cfg
>   https://github.com/yjxiong/darknet/pull/1/files#diff-15 (119)
> - _M_ pyDarknet/**init**.py
>   https://github.com/yjxiong/darknet/pull/1/files#diff-16 (18)
> - _M_ src/activation_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-17 (20)
> - _A_ src/activation_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-18 (58)
> - _A_ src/activation_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-19 (19)
> - _M_ src/activations.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-20 (14)
> - _M_ src/activations.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-21 (10)
> - _M_ src/avgpool_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-22 (3)
> - _M_ src/avgpool_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-23 (2)
> - _M_ src/avgpool_layer_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-24 (4)
> - _M_ src/blas.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-25 (88)
> - _M_ src/blas.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-26 (30)
> - _M_ src/blas_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-27 (328)
> - _M_ src/box.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-28 (47)
> - _M_ src/box.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-29 (1)
> - _A_ src/classifier.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-30 (466)
> - _M_ src/coco.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-31 (487)
> - _A_ src/coco_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-32 (151)
> - _M_ src/col2im_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-33 (4)
> - _M_ src/compare.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-34 (110)
> - _M_ src/connected_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-35 (181)
> - _M_ src/connected_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-36 (4)
> - _M_ src/convolutional_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-37 (200)
> - _M_ src/convolutional_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-38 (193)
> - _M_ src/convolutional_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-39 (9)
> - _M_ src/cost_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-40 (21)
> - _M_ src/cost_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-41 (2)
> - _M_ src/crop_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-42 (38)
> - _M_ src/crop_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-43 (3)
> - _M_ src/crop_layer_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-44 (25)
> - _M_ src/cuda.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-45 (1)
> - _M_ src/darknet.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-46 (379)
> - _M_ src/data.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-47 (124)
> - _M_ src/data.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-48 (4)
> - _M_ src/deconvolutional_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-49 (7)
> - _M_ src/deconvolutional_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-50 (3)
> - _M_ src/deconvolutional_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-51 (2)
> - _M_ src/detection_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-52 (294)
> - _M_ src/detection_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-53 (10)
> - _M_ src/dice.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-54 (2)
> - _M_ src/dropout_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-55 (3)
> - _M_ src/dropout_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-56 (2)
> - _M_ src/dropout_layer_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-57 (5)
> - _M_ src/im2col_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-58 (4)
> - _M_ src/image.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-59 (124)
> - _M_ src/image.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-60 (46)
> - _M_ src/imagenet.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-61 (16)
> - _M_ src/layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-62 (81)
> - _A_ src/local_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-63 (275)
> - _A_ src/local_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-64 (31)
> - _M_ src/maxpool_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-65 (1)
> - _M_ src/maxpool_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-66 (2)
> - _M_ src/maxpool_layer_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-67 (4)
> - _M_ src/network.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-68 (73)
> - _M_ src/network.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-69 (78)
> - _M_ src/network_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-70 (44)
> - _M_ src/nightmare.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-71 (124)
> - _M_ src/normalization_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-72 (2)
> - _M_ src/option_list.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-73 (18)
> - _M_ src/option_list.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-74 (1)
> - _M_ src/parser.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-75 (286)
> - _M_ src/parser.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-76 (16)
> - _M_ src/pydarknet.cpp
>   https://github.com/yjxiong/darknet/pull/1/files#diff-77 (328)
> - _A_ src/region_layer.cxx
>   https://github.com/yjxiong/darknet/pull/1/files#diff-78 (194)
> - _A_ src/region_layer.hxx
>   https://github.com/yjxiong/darknet/pull/1/files#diff-79 (18)
> - _A_ src/rnn.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-80 (0)
> - _A_ src/rnn_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-81 (0)
> - _A_ src/rnn_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-82 (0)
> - _A_ src/shortcut_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-83 (0)
> - _A_ src/shortcut_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-84 (0)
> - _I_ src/softmax_layer.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-85 (0)
> - _I_ src/softmax_layer.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-86 (0)
> - _I_ src/softmax_layer_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-87 (0)
> - _I_ src/swag.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-88 (0)
> - _I_ src/utils.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-89 (0)
> - _I_ src/utils.h
>   https://github.com/yjxiong/darknet/pull/1/files#diff-90 (0)
> - _I_ src/yolo.c
>   https://github.com/yjxiong/darknet/pull/1/files#diff-91 (0)
> - _A_ src/yolo_kernels.cu
>   https://github.com/yjxiong/darknet/pull/1/files#diff-92 (0)
> 
> Patch Links:
> - https://github.com/yjxiong/darknet/pull/1.patch
> - https://github.com/yjxiong/darknet/pull/1.diff
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/yjxiong/darknet/pull/1.

## 

Best regards,

Yuanjun

File: CMakeLists.txt [hunk 1]
```diff
@@ -28,32 +28,32 @@ else()
...
 FIND_PACKAGE(PythonLibs REQUIRED)
+FIND_PACKAGE(OpenCV REQUIRED)
+
 include_directories(SYSTEM ${Boost_INCLUDE_DIR})
 include_directories(SYSTEM ${PYTHON_INCLUDE_DIR})
+include_directories(SYSTEM ${OpenCV_INCLUDE_DIR})
 
 include_directories(src)
-set(SRC_DIR src)
+set(SRC_DIR /set/path/pydarknet)
```
Comment (yjxiong) [inline CMakeLists.txt]: If you do need to change this variable, better change it in a CMake's fasion, e.g. setting the variable in `cmake` command.

File: Makefile [hunk 2]
```diff
@@ -2,7 +2,7 @@ GPU=1
 OPENCV=1
 DEBUG=0
 
-ARCH= --gpu-architecture=compute_30 --gpu-code=compute_30
+ARCH= --gpu-architecture=compute_20 --gpu-code=compute_20
```
Comment (yjxiong) [inline Makefile]: No need to change the arch number here as we are not using the original Makefile.

File: README.md [hunk 3]
```diff
@@ -6,7 +6,7 @@ This repo is for the python interface of YOLO object detector.
 
 
 #Build
-To build, run
+To build, change (CMakeLists.txt:40) to your path, run
```
Comment (yjxiong) [inline README.md]: Same as the first comment.

**Comment by yjxiong** [comment]
Hi @bongmo , I have read through the codes. I made some minor comments. Please see if you have to some time to look at them. Then we should be good to merge.

Thanks.

**Comment by bongmo** [comment]
Hi, @yjxiong , Thanks for your warm review.
I'll change all parts following to your recommendation.

Thanks.

**Comment by yjxiong** [comment]
Hi @bongmo , I got some compilation errors on my desktop. Did you succeed in the compilation?

The log is as follows

```
[  1%] Building C object CMakeFiles/dnl.dir/src/region_layer.c.o
In file included from /home/alex/darknet/src/softmax_layer.h:4:0,
                 from /home/alex/darknet/src/region_layer.c:3:
/home/alex/darknet/src/network.h:53:3: error: conflicting types for ‘network_state’
 } network_state;
   ^
In file included from /home/alex/darknet/src/region_layer.h:4:0,
                 from /home/alex/darknet/src/region_layer.c:1:
/home/alex/darknet/src/params.h:9:3: note: previous declaration of ‘network_state’ was here
 } network_state;
   ^
/home/alex/darknet/src/region_layer.c: In function ‘make_region_layer’:
/home/alex/darknet/src/region_layer.c:16:14: error: ‘REGION’ undeclared (first use in this function)
     l.type = REGION;
              ^
/home/alex/darknet/src/region_layer.c:16:14: note: each undeclared identifier is reported only once for each function it appears in
/home/alex/darknet/src/region_layer.c: At top level:
/home/alex/darknet/src/region_layer.c:42:6: error: conflicting types for ‘forward_region_layer’
 void forward_region_layer(const region_layer l, network_state state)
      ^
In file included from /home/alex/darknet/src/region_layer.c:1:0:
/home/alex/darknet/src/region_layer.h:10:6: note: previous declaration of ‘forward_region_layer’ was here
 void forward_region_layer(const region_layer l, network_state state);
      ^
/home/alex/darknet/src/region_layer.c: In function ‘forward_region_layer’:
/home/alex/darknet/src/region_layer.c:53:17: error: incompatible type for argument 3 of ‘softmax_array’
                 softmax_array(l.output + index + offset, l.classes,
                 ^
In file included from /home/alex/darknet/src/region_layer.c:3:0:
/home/alex/darknet/src/softmax_layer.h:8:6: note: expected ‘float’ but argument is of type ‘float * const’
 void softmax_array(float *input, int n, float temp, float *output);
      ^
/home/alex/darknet/src/region_layer.c:53:17: error: too few arguments to function ‘softmax_array’
                 softmax_array(l.output + index + offset, l.classes,
                 ^
In file included from /home/alex/darknet/src/region_layer.c:3:0:
/home/alex/darknet/src/softmax_layer.h:8:6: note: declared here
 void softmax_array(float *input, int n, float temp, float *output);
      ^
/home/alex/darknet/src/region_layer.c: At top level:
/home/alex/darknet/src/region_layer.c:160:6: error: conflicting types for ‘backward_region_layer’
 void backward_region_layer(const region_layer l, network_state state)
      ^
In file included from /home/alex/darknet/src/region_layer.c:1:0:
/home/alex/darknet/src/region_layer.h:11:6: note: previous declaration of ‘backward_region_layer’ was here
 void backward_region_layer(const region_layer l, network_state state);
      ^
In file included from /home/alex/darknet/src/region_layer.c:1:0:
/home/alex/darknet/src/region_layer.h:11:6: note: previous declaration of ‘backward_region_layer’ was here
 void backward_region_layer(const region_layer l, network_state state);
      ^
/home/alex/darknet/src/region_layer.c:167:6: error: conflicting types for ‘forward_region_layer_gpu’
 void forward_region_layer_gpu(const region_layer l, network_state state)
      ^
In file included from /home/alex/darknet/src/region_layer.c:1:0:
/home/alex/darknet/src/region_layer.h:14:6: note: previous declaration of ‘forward_region_layer_gpu’ was here
 void forward_region_layer_gpu(const region_layer l, network_state state);
      ^
/home/alex/darknet/src/region_layer.c:188:6: error: conflicting types for ‘backward_region_layer_gpu’
 void backward_region_layer_gpu(region_layer l, network_state state)
      ^
In file included from /home/alex/darknet/src/region_layer.c:1:0:
/home/alex/darknet/src/region_layer.h:15:6: note: previous declaration of ‘backward_region_layer_gpu’ was here
 void backward_region_layer_gpu(region_layer l, network_state state);
      ^
make[2]: *** [CMakeFiles/dnl.dir/src/region_layer.c.o] Error 1
make[1]: *** [CMakeFiles/dnl.dir/all] Error 2
make: *** [all] Error 2

```

**Comment by bongmo** [comment]
Hi, @yjxiong , Sorry to late response.

I found errors from invalid source code named "region_layer.c".
So I remove the code and can build success.

Thanks.

**Comment by yjxiong** [comment]
Thanks! I will look at it asap.

**Comment by yjxiong** [comment]
Merged, thanks a lot.
